File: src-architecture/game/creator/SolarSystemCreator.java
package game.creator;

import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import javax.media.opengl.GL2;

import model.Coords;
import model.celestials.CelestialBody;
import model.celestials.Star;
import model.ship.Mesh;
import model.ship.Ship;
import model.ship.parts.Engine;
import model.ship.parts.Position;
import model.system.StarSystem;
import model.system.parts.SkyBox;
import engine.utils.JsonSystemParser;
import game.architecture.Constants;
import game.architecture.GameComponentContainer;

/**
 * Creates Solar system.
 */
public class SolarSystemCreator extends SystemCreator {

    /* ========== PUBLIC ========== */
    /**
     * Method that creates Solar system.
     * 
     * TODO In future - systems should be loaded from JSON file.
     */
    public static StarSystem create(GL2 gl) throws IOException {

        List<CelestialBody> bodies = new LinkedList<CelestialBody>();

        Star sun = createStar(gl, bodies);
        createCelestialBodies(gl, bodies, sun);
        SkyBox skyBox = createSkyBox("solarSystem");

        ArrayList<Ship> ships = new ArrayList<Ship>();

        // Mesh exxonDrawer = GameComponentContainer.meshRepository.getById(0);
        Mesh exxonDrawer = GameComponentContainer.meshRepository.getById(0);

        ships.add(new Ship(new Position(new Coords(Constants.AU_PARAMETER, 0, 0), 90, 0), new Engine(2), exxonDrawer, 1));

        GameComponentContainer.shipRepository.add(ships);

        StarSystem system = new StarSystem(skyBox, ships, bodies);
        JsonSystemParser.saveStarSystem("solarSystem", system);

        return system;
    }

    /* ========== PRIVATE ========== */
    private static Star createStar(GL2 gl, List<CelestialBody> celestialBodies) throws IOException {
        return star(gl, "Sun", 190671, 1973.20, celestialBodies);
    }

    private static void createCelestialBodies(GL2 gl, List<CelestialBody> bodies, Star sun) throws IOException {
        planet(gl, "Mercury", null, null, 2395, 10.892, 0.03, sun, 0.3871, 47.87, 120, bodies);
        planet(gl, "VenusClouds", null, null, 6023, -6.52, 2.64, sun, 0.7233, 35.02, 60, bodies);

        CelestialBody earth = planet(gl, "Earth", "EarthClouds", null, 6230, 1601.21, 23.44, sun, 1.0000, 29.78, 60,
                bodies);
        moon(gl, "Moon", null, 1738, 16.65, 1.54, earth, 0.000256, 1.022, 0, bodies);

        planet(gl, "Mars", null, null, 3402, 868.22, 25.19, sun, 1.5236, 24.16, 100, bodies);

        CelestialBody jupiter = planet(gl, "Jupiter", null, null, 71095, 45360.00, 3.12, sun, 5.2043, 17.82, 130,
                bodies);
        moon(gl, "Io", null, 1821, 15.00, 1.00, jupiter, 0.000281, 13.07, 100, bodies);
        moon(gl, "Europe", null, 1561, 15.00, 1.00, jupiter, 0.000447, 1.500, 80, bodies);
        moon(gl, "Ganymede", null, 2631, 15.00, 1.00, jupiter, 0.000714, 1.500, 100, bodies);
        moon(gl, "Callisto", null, 2410, 15.00, 1.00, jupiter, 0.001255, 1.500, 100, bodies);

        CelestialBody saturn = planet(gl, "Saturn", null, "SaturnRings", 64095, 35500.00, 26.73, sun, 9.5370, 9.63, 21,
                bodies);
        moon(gl, "Titan", null, 2575, 15.00, 0.31, saturn, 0.000815, 1.500, 51, bodies);
        moon(gl, "Rhea", null, 764, 15.00, 0.33, saturn, 0.000351, 1.500, 10, bodies);
        moon(gl, "Iapetus", null, 718, 15.00, 8.30, saturn, 0.002374, 1.500, 120, bodies);
        moon(gl, "Dione", null, 559, 15.00, 0.03, saturn, 0.000252, 1.500, 70, bodies);
        moon(gl, "Tethys", null, 529, 15.00, 1.09, saturn, 0.000291, 1.500, 70, bodies);

        CelestialBody uranus = planet(gl, "Uranus", null, null, 41095, 9320.00, 97.77, sun, 12.1912, 6.79, 100, bodies);
        moon(gl, "Titania", null, 788, 15.00, 1.09, uranus, 0.000291, 1.500, 70, bodies);
        moon(gl, "Oberon", null, 761, 15.00, 1.09, uranus, 0.000389, 1.500, 170, bodies);
        moon(gl, "Umbriel", null, 584, 15.00, 1.09, uranus, 0.000177, 1.500, 270, bodies);
        moon(gl, "Ariel", null, 578, 15.00, 1.09, uranus, 0.000127, 1.500, 80, bodies);

        CelestialBody neptune = planet(gl, "Neptune", null, null, 40095, 9275.52, 15.58, sun, 14.0689, 5.48, 0, bodies);
        moon(gl, "Triton", null, 1353, 15.00, 156.80, neptune, 0.000256, -4.75, 70, bodies);

        wormhole(gl, "BlueGalaxy", 10000.00, 10421.43, 0, sun, 4.6432, 10.32, 180, 0, 1, bodies);
    }
}

File: src-architecture/game/creator/SolarSystemCreator.kt
package game.creator

import engine.math.Degree
import game.architecture.GameComponentContainer
import model.Coords
import model.celestials.CelestialBody
import model.celestials.Star
import model.ship.Ship
import model.ship.parts.Engine
import model.ship.parts.Position
import model.system.StarSystem
import java.util.*

class SolarSystemCreator : SystemCreator() {

    fun create(): StarSystem {

        val bodies = LinkedList<CelestialBody>()

        val sun = createStar(bodies)
        createCelestialBodies(bodies, sun)
        val skyBox = createSkyBox("solarSystem")

        val ships = ArrayList<Ship>()

        val exxonDrawer = GameComponentContainer.meshRepository.getById(0)

        ships.add(
                Ship(
                        Position(
                                Coords(-950.0, 0.0, 0.0),
                                Degree(90f),
                                Degree(0f)),
                        Engine(2f, 0f, 0f, Degree.ZERO, Degree.ZERO),
                        exxonDrawer, 10.0))

        GameComponentContainer.shipRepository.add(ships)

        return StarSystem(skyBox, ships, bodies)
    }

    private fun createStar(celestialBodies: MutableList<CelestialBody>): Star {
        return star("Sun", 190671.0F, 1973.20f, celestialBodies)
    }

    private fun createCelestialBodies(bodies: MutableList<CelestialBody>, sun: Star) {
        planet("Mercury", null, null, 2395f, 10.892f, 0.03, sun, 0.3871f, 47.87f, 120f, bodies)
        planet("VenusClouds", null, null, 6023f, -6.52f, 2.64, sun, 0.7233f, 35.02f, 60f, bodies)

        val earth = planet("Earth", "EarthClouds", null, 6230f, 1601.21f, 23.44, sun, 1.0000f, 29.78f, 60f,
                bodies)
        moon("Moon", 1738.0F, 16.65f, 1.54, earth, 0.00256f, 1.022f, 0f, bodies)

        planet("Mars", null, null, 3402f, 868.22f, 25.19, sun, 1.5236f, 24.16f, 100f, bodies)

        val jupiter = planet("Jupiter", null, null, 71095f, 45360.00f, 3.12, sun, 5.2043f, 17.82f, 130f,
                bodies)
        moon("Io", 1821.0F, 15.00f, 1.00, jupiter, 0.00281f, 13.07f, 100f, bodies)
        moon("Europe", 1561.0F, 15.00f, 1.00, jupiter, 0.00447f, 1.500f, 80f, bodies)
        moon("Ganymede", 2631.0F, 15.00f, 1.00, jupiter, 0.00714f, 1.500f, 100f, bodies)
        moon("Callisto", 2410.0F, 15.00f, 1.00, jupiter, 0.01255f, 1.500f, 100f, bodies)

        val saturn = planet("Saturn", null, "SaturnRings", 64095f, 35500.00f, 26.73, sun, 9.5370f, 9.63f, 21f,
                bodies)
        moon("Titan", 2575.0F, 15.00f, 0.31, saturn, 0.00815f, 1.500f, 51f, bodies)
        moon("Rhea", 764.0F, 15.00f, 0.33, saturn, 0.00351f, 1.500f, 10f, bodies)
        moon("Iapetus", 718.0F, 15.00f, 8.30, saturn, 0.02374f, 1.500f, 120f, bodies)
        moon("Dione", 559.0F, 15.00f, 0.03, saturn, 0.00252f, 1.500f, 70f, bodies)
        moon("Tethys", 529.0F, 15.00f, 1.09, saturn, 0.00291f, 1.500f, 70f, bodies)

        val uranus = planet("Uranus", null, null, 41095f, 9320.00f, 97.77, sun, 12.1912f, 6.79f, 100f, bodies)
        moon("Titania", 788.0F, 15.00f, 1.09, uranus, 0.00291f, 1.500f, 70f, bodies)
        moon("Oberon", 761.0F, 15.00f, 1.09, uranus, 0.00389f, 1.500f, 170f, bodies)
        moon("Umbriel", 584.0F, 15.00f, 1.09, uranus, 0.00177f, 1.500f, 270f, bodies)
        moon("Ariel", 578.0F, 15.00f, 1.09, uranus, 0.00127f, 1.500f, 80f, bodies)

        val neptune = planet("Neptune", null, null, 40095f, 9275.52f, 15.58, sun, 14.0689f, 5.48f, 0f, bodies)
        moon("Triton", 1353.0F, 15.00f, 156.80, neptune, 0.00256f, -4.75f, 70f, bodies)

        wormhole("BlueGalaxy", 10000.00F, 10421.43f, 0.0, sun, 4.6432f, 10.32f, 180f, 0, 1, bodies)
    }
}